sigstore/oidc.py

```diff
@@ -0,0 +1,184 @@
+    """
+    Retrieve an OpenID Connect token from the Sigstore provider
+
+    This function and the components that it relies on are based off of:
+    https://github.com/psteniusubi/python-sample
+    """
+    from sigstore._internal.oidc.oauth import OAuthFlow
+
+    force_oob = os.getenv("SIGSTORE_OAUTH_FORCE_OOB") is not None
+
+    code: str
+    with OAuthFlow(client_id, client_secret, issuer) as server:
+        # Launch web browser
+        if not force_oob and webbrowser.open(server.base_uri):
+            print("Waiting for browser interaction...")
+        else:
+            server.enable_oob()
+            print(f"Go to the following link in a browser:\n\n\t{server.auth_endpoint}")
+
+        if not server.is_oob():
+            # Wait until the redirect server populates the response
+            while server.auth_response is None:
+                time.sleep(0.1)
+
+            auth_error = server.auth_response.get("error")
+            if auth_error is not None:
+                raise IdentityError(
+                    f"Error response from auth endpoint: {auth_error[0]}"
+                )
+            code = server.auth_response["code"][0]
+        else:
+            # In the out-of-band case, we wait until the user provides the code
+            code = input("Enter verification code: ")
+
+    # Provide code to token endpoint
+    data = {
+        "grant_type": "authorization_code",
+        "redirect_uri": server.redirect_uri,
+        "code": code,
+        "code_verifier": server.oauth_session.code_verifier,
+    }
+    auth = (
+        client_id,
+        client_secret,
+    )
+    logging.debug(f"PAYLOAD: data={data}")
+    resp: requests.Response = requests.post(
+        issuer.token_endpoint,
+        data=data,
+        auth=auth,
+    )
+
+    try:
+        resp.raise_for_status()
+    except requests.HTTPError as http_error:
+        raise IdentityError from http_error
+
+    token_json = resp.json()
+    token_error = token_json.get("error")
+    if token_error is not None:
+        raise IdentityError(f"Error response from token endpoint: {token_error}")
+
+    return str(token_json["access_token"])
+
+
+class AmbientCredentialError(IdentityError):
+    """
+    Raised when an ambient credential should be present, but
+    can't be retrieved (e.g. network failure).
+    """
+
+    pass
+
+
+class GitHubOidcPermissionCredentialError(AmbientCredentialError):
+    """
+    Raised when the current GitHub Actions environment doesn't have permission
+    to retrieve an OIDC token.
+    """
+
+    pass
+
+
+def detect_credential() -> Optional[str]:
+    """
+    Try each ambient credential detector, returning the first one to succeed
+    or `None` if all fail.
+
+    Raises `AmbientCredentialError` if any detector fails internally (i.e.
+    detects a credential, but cannot retrieve it).
+    """
+    from sigstore._internal.oidc.ambient import detect_gcp, detect_github
```

Comment:
Had to do this to avoid circular imports. I figured this would be better than further complicating the tree with more modules, but I can try fixing this another way if you'd like.